Tests/DatadogTests/Datadog/URLSessionAutoInstrumentation/Interception/AppStateListenerTests.swift

```diff
@@ -0,0 +1,197 @@
+    func testWhenAppResignActiveAndBecomeActive_thenAppStateHistoryIsRecorded() {
+        let startDate = Date.mockDecember15th2019At10AMUTC()
+        let listener = AppStateListener(
+            dateProvider: RelativeDateProvider(startingFrom: startDate, advancingBySeconds: 1.0)
+        )
+
+        NotificationCenter.default.post(name: UIApplication.willResignActiveNotification, object: nil)
+        NotificationCenter.default.post(name: UIApplication.didBecomeActiveNotification, object: nil)
+
+        let expected = AppStateHistory(
+            initialState: .init(isActive: true, date: startDate),
+            changes: [
+                .init(isActive: false, date: startDate + 1.0),
+                .init(isActive: true, date: startDate + 2.0)
+            ],
+            finalDate: startDate + 3.0
+        )
+        XCTAssertEqual(listener.history, expected)
+    }
+
+    func testWhenAppBecomeActiveAndResignActive_thenAppStateHistoryIsRecorded() {
+        let startDate = Date.mockDecember15th2019At10AMUTC()
+        let listener = AppStateListener(
+            dateProvider: RelativeDateProvider(startingFrom: startDate, advancingBySeconds: 1.0)
+        )
+        NotificationCenter.default.post(name: UIApplication.didBecomeActiveNotification, object: nil)
+        NotificationCenter.default.post(name: UIApplication.willResignActiveNotification, object: nil)
+
+        let expected = AppStateHistory(
+            initialState: .init(isActive: true, date: startDate),
+            changes: [
+                .init(isActive: true, date: startDate + 1.0),
+                .init(isActive: false, date: startDate + 2.0)
+            ],
+            finalDate: startDate + 3.0
+        )
+        XCTAssertEqual(listener.history, expected)
+    }
+
+    func testWhenAppStateHistoryIsRetrieved_thenFinalDateOfHistoryChanges() {
+        let startDate = Date.mockDecember15th2019At10AMUTC()
+        let listener = AppStateListener(
+            dateProvider: RelativeDateProvider(startingFrom: startDate, advancingBySeconds: 1.0)
+        )
+        let history1 = listener.history
+        let history2 = listener.history
+
+        XCTAssertEqual(history2.finalState.date.timeIntervalSince(history1.finalState.date), 1.0)
+    }
+
+    func testWhenAppStateListenerIsCalledFromDifferentThreads_thenItWorks() {
+        let listener = AppStateListener(dateProvider: SystemDateProvider())
+        DispatchQueue.concurrentPerform(iterations: 10_000) { iteration in
+            // write
+            if iteration < 1_000 {
+                let nc = NotificationCenter.default
+                nc.post(
+                    name: (Bool.random() ?
+                            UIApplication.willResignActiveNotification :
+                            UIApplication.didBecomeActiveNotification),
+                    object: nil
+                )
+            }
+            // read
+            XCTAssertFalse(listener.history.changes.isEmpty)
```

Comment:
We can use existing [`callConcurrently {}` helper](https://github.com/DataDog/dd-sdk-ios/blob/36c57c0ea5e9323be92b4a87d903350aa33536cb/Tests/DatadogTests/Helpers/XCTestCase.swift#L27-L37) to mix concurrent writes and reads uniformly:

```swift
// swiftlint:disable opening_brace
callConcurrently(
   closures: [
      { /* write (post notification) */ }
      { /* read (access value) */ }
   ],
   iterations: 1_000
)
// swiftlint:enable opening_brace
```

This should give us approximately 50% of overlap between writes and reads. Current solution is only overlapping 10% (1K out of 10K).

Reply:
as you said before, the real use case is write-rarely&read-often.
this test case is closer to the real use case.